File: Scripts/stable_regions/defining_stable_regions.R
cat("Running: defining_stable_regions.R\n")

# Installing and Loading Packages
required_packages <- c("dplyr", "stringr", "readr", "ggplot2", "tidyr", "zoo", "pastecs")

not_installed <- required_packages[!(required_packages %in% installed.packages()[ , "Package"])]   

# Install missing packages
if(length(not_installed)) install.packages(not_installed)    

# Load all required packages
invisible(lapply(required_packages, require, character.only = TRUE))

# Cleaning Environment
rm(list = ls())

################################################################################

# Creating Directories
if(!dir.exists(file.path("Output", "stable_regions"))) dir.create(file.path("Output", "stable_regions"))

################################################################################

# Importing Data
filtered_df <- read.csv(file.path("Output", "thermodynamics", "foldx_stability_filtered.csv"))

# Importing the mean delta G per residue for each PDB 
PDB_mean_per_res <- read.csv(file.path("Output", "thermodynamics", "mean_deltaG_per_residue.csv"))

# Playing with different sliding windows

window_size <- 3

PDB_mean_per_res <- PDB_mean_per_res %>%
                      group_by(pdb_id) %>%
                        mutate(sliding_window = rollapply(mean_energy, width = window_size, FUN = mean, align = "center", fill = NA)) %>% 
                          ungroup()

################################################################################

# Finding local minima (troughs)
mean_at_each_pos <- PDB_mean_per_res %>% group_by(Pos) %>% summarise(mean_Pos_sliding_window = mean(sliding_window, na.rm = TRUE))
mean_at_each_pos <- na.omit(mean_at_each_pos)


mean_stability <- mean(PDB_mean_per_res$sliding_window, na.rm = TRUE)

# Get the turning points object
tp <- pastecs::turnpoints(mean_at_each_pos$mean_Pos_sliding_window, calc.proba = TRUE)

# Creating a turnpoint df from tp
info <- extract(tp, peak = 1, pit = -1)
type <- info[info != 0]

tp_df <- data.frame(Pos = mean_at_each_pos$Pos[tp$tppos], # tppos counts each data point, since my data doesnt start at 1 I need to account for it
                    info = as.character(type),
                    probability = tp$proba)

# Define a probability threshold
threshold <- 0.15  # for instance, if you're interested in low probability values

# Filter the turning points object based on probability
tp_df <- tp_df %>% filter(probability < threshold)

# Create a data frame for plotting
mean_at_each_pos <- merge(mean_at_each_pos, tp_df, by = "Pos", all = TRUE)
mean_at_each_pos$info[is.na(mean_at_each_pos$info)] <- "0"

#####################################################################################################

### Selecting Stable Regions ###

# Selecting peaks above the mean line
significant_peaks <- mean_at_each_pos$Pos[mean_at_each_pos$info == 1]

# Initialize variables
stabilisers <- c()
region_list <- c()
region <- 1
stabiliser_count <- 0

# Loop through the data to find stabilizers
for (i in seq_along(mean_at_each_pos$mean_Pos_sliding_window)) {
  
  # Check if the current position is stabilizing
  if (mean_at_each_pos$mean_Pos_sliding_window[i] < mean_stability & mean_at_each_pos$info[i] != 1) {
    stabilisers <- c(stabilisers, mean_at_each_pos$Pos[i])
    region_list <- c(region_list, region)
    stabiliser_count <- stabiliser_count + 1
  }
  
  # If a significant peak is hit, check if the current region has more than 1 stabilizer
  if (mean_at_each_pos$Pos[i] %in% significant_peaks) {
    
    # Only proceed if there are more than 1 stabilizers in the region
    if (stabiliser_count > 1) {
      region <- region + 1
    } else {
      # Remove stabilisers with less than 2 occurrences
      stabilisers <- stabilisers[region_list != region]
      region_list <- region_list[region_list != region]
    }
    
    # Reset stabilizer count for the next region
    stabiliser_count <- 0
  }
}

# Creating a data frame of only significantly stabilising positions and which group
stabilising_positions <- data.frame(Pos = stabilisers,
                          region = region_list)

# Fixing the group numbers to start from 1
unique_regions <- unique(region_list)

for (i in seq_along(unique_regions)) {
  
  stabilising_positions$region[stabilising_positions$region == unique_regions[i]] <- i
  
}

# Summarising into stable regions
stable_regions <- stabilising_positions %>% 
                    group_by(region) %>% 
                      summarise(min = min(Pos), max = max(Pos))

stable_regions <- stable_regions %>% mutate(residues = paste0(min, "-", max))

stable_regions$region <- as.character(stable_regions$region)

stable_regions$region <- factor(stable_regions$region, levels = sort(as.numeric(unique(stable_regions$region))))

# Assigning colours

colours <- c("red", "blue", "gold2", "purple", "darkorange2", "cyan", "green", "pink", "chocolate4", 
                    "darkgreen", "magenta", "deepskyblue", "darkviolet", "coral", "darkslategray")

# Extend the color vector to match `region`
region_colours <- rep(colours, length.out = region)


# Visualising Stabilising Regions
ggplot(mean_at_each_pos, aes(x = Pos, y = mean_Pos_sliding_window)) +
  geom_rect(data = stable_regions, aes(xmin = min - 0.5, xmax = max + 0.5, ymin = -Inf, ymax = Inf, fill = region),
            alpha = 0.75, inherit.aes = FALSE)+#, fill = "grey50") +
  geom_line(color = "black", linewidth = 1) +
  geom_point(aes(color = info), size = 2) +
  scale_fill_manual(values = region_colours) +
  scale_color_manual(values = c("1" = "blue", "0" = "black", "-1" = "red"),
                     label= c("Trough", "NA", "Peak")) +
  scale_x_continuous(breaks = seq(0,1000,10)) +
  geom_hline(yintercept = mean_stability, linetype = "solid", size = 1) +  # Mean line
  labs(title = "Turning Points", x = "Residue Number", y = "Mean Pos Sliding Window", colour = "Type", fill = "Region") +
  theme_minimal() +
  theme(panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
        panel.grid.major = element_line(colour = "grey", linewidth = 0.25, linetype = "dashed"),
        panel.grid.minor = element_line(colour = "grey", linewidth = 0.25, linetype = "dashed"),
        plot.title = element_text(size = 14, colour = "black", face = "bold", hjust = 0.5),
        axis.text = element_text(size = 10, colour = "black", face = "bold"),
        axis.title = element_text(size = 12, face = "bold"),
        legend.title = element_text(size = 12, face = "bold", colour = "black"),
        legend.text = element_text(size = 10, colour = "black")) 

ggsave(plot = last_plot(), file = file.path("Output", "stable_regions", "turning_points_and_regions.png"), height = 5, width = 8, bg = "white")

ggplot(data = PDB_mean_per_res, aes(x=Pos)) +
  geom_rect(data = stable_regions, aes(xmin = min - 0.5, xmax = max + 0.5, ymin = -Inf, ymax = Inf, fill = region),
            alpha = 0.75, inherit.aes = FALSE) +
  scale_fill_manual(values = region_colours) +
  geom_hline(yintercept = 0, colour = "black", linewidth = 1, linetype = "solid", alpha = 0.9) +
  geom_line(aes(y = sliding_window, group = interaction(pdb_id, cumsum(c(0, diff(Pos) != 1)))), 
                 linewidth = 0.5, alpha = 0.25) +
  ggtitle(paste0("Sliding Window of ", window_size)) +
  ylab(bquote(bold(Delta * "G"*degree * " per residue (kcal.mol"^-1 * ")"))) +
  xlab("Residue Number") +
  scale_x_continuous(breaks = seq(0,99999, by = 10)) +
  theme_classic() +
  theme(panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
        panel.grid.major = element_line(colour = "grey", linewidth = 0.25, linetype = "dashed"),
        panel.grid.minor = element_line(colour = "grey", linewidth = 0.25, linetype = "dashed"),
        plot.title = element_text(size = 14, colour = "black", face = "bold", hjust = 0.5),
        axis.text = element_text(size = 10, colour = "black", face = "bold"),
        axis.title = element_text(size = 12, face = "bold"),
        legend.position = "none")

ggsave(plot = last_plot(), file = file.path("Output", "stable_regions", "sliding_window_and_regions.png"), height = 4, width = 8, bg = "white")


ggplot(data = PDB_mean_per_res, aes(x=Pos)) +
  geom_rect(data = stable_regions, aes(xmin = min - 0.5, xmax = max + 0.5, ymin = -Inf, ymax = Inf, fill = region),
            alpha = 0.75, inherit.aes = FALSE) +
  scale_fill_manual(values = region_colours) +
  geom_hline(yintercept = 0, colour = "black", linewidth = 1, linetype = "solid", alpha = 0.9) +
  geom_line(aes(y = mean_energy, group = interaction(pdb_id, cumsum(c(0, diff(Pos) != 1)))), 
            linewidth = 0.5, alpha = 0.25) +
  ylab(bquote(bold(Delta * "G"*degree * " per residue (kcal.mol"^-1 * ")"))) +
  xlab("Residue Number") +
  scale_x_continuous(breaks = seq(0,99999, by = 10)) +
  theme_classic() +
  theme(panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
        panel.grid.major = element_line(colour = "grey", linewidth = 0.25, linetype = "dashed"),
        panel.grid.minor = element_line(colour = "grey", linewidth = 0.25, linetype = "dashed"),
        axis.text = element_text(size = 10, colour = "black", face = "bold"),
        axis.title = element_text(size = 12, face = "bold"),
        legend.position = "none")

ggsave(plot = last_plot(), file = file.path("Output", "stable_regions", "raw_delta_G_and_regions.png"), height = 4, width = 8, bg = "white")

# Saving stable regions
stable_regions_minimal <- stable_regions %>% select("residues", "region")
write.csv(stable_regions_minimal, file = file.path("Output", "stable_regions", "stable_regions.csv"), row.names = FALSE)

File: Scripts/stable_regions/defining_stable_regions.py
import pandas as pd
import numpy as np
import os
import sys
from scipy.signal import find_peaks
from plotnine import *
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

#############
### Setup ###
#############

# Creating output_dir
output_dir = os.path.join("Output", "stable_regions")
if not os.path.exists(output_dir):
    os.makedirs(output_dir)
    
# Importing data
filtered_df = pd.read_csv(os.path.join("Output", "thermodynamics", "foldx_stability_filtered.csv"))
mean_dG_per_res_by_PDB = pd.read_csv(os.path.join("Output", "thermodynamics", "mean_deltaG_per_residue.csv"))

# Read window_size argument 
if len(sys.argv) > 1: 
    try: 
        window_size = int(sys.argv[1]) 
        print(f"Window size: {window_size}") 
    except ValueError: 
        print("Warning: invalid window size provided — defaulting to 3.") 
        window_size = 3 
else: 
    window_size = 3 
    print("No window size provided, defaulting to 3")

################
### Analysis ###
################

# Calculating the rolling average 
mean_dG_per_res_by_PDB["sliding_window"] = (
    mean_dG_per_res_by_PDB
    .groupby("pdb_id")["mean_energy"]
    .transform(lambda x: x.rolling(window = window_size, center = True, min_periods = window_size).mean())
)

# Calculating the mean dG per residue across all PDBs
mean_dG_by_pos = (
    mean_dG_per_res_by_PDB
    .groupby("Pos", as_index = False)["sliding_window"]
    .mean()
    .dropna()
)

# Calculating the mean stability
mean_stability = mean_dG_per_res_by_PDB["sliding_window"].mean()
#mean_stability = mean_dG_per_res_by_PDB["mean_energy"].mean()

#######################################
### Finding Local Minima and Maxima ###
#######################################

x = mean_dG_by_pos["Pos"]
y = mean_dG_by_pos["sliding_window"]

# Find local maxima (peaks)
peaks, peak_props = find_peaks(y, prominence = 0.01)  
# Find local minima (by inverting the signal)
pits, pit_props = find_peaks(-y, prominence = 0.01)

# Combine into a single dataframe of turning points
tp_df = pd.concat([
    pd.DataFrame({
        "Pos": x.iloc[peaks].values,
        "type": "peak",
        "prominence": peak_props["prominences"]
    }),
    pd.DataFrame({
        "Pos": x.iloc[pits].values,
        "type": "pit",
        "prominence": pit_props["prominences"]
    })
])

# Sort by position
tp_df = tp_df.sort_values("Pos").reset_index(drop = True)

# filtering turning points by prominence
threshold = 0.15  
tp_df = tp_df[tp_df["prominence"] > threshold]

# Adding in sliding_window to tp_df
tp_df = pd.merge(tp_df, mean_dG_by_pos, on = "Pos", how = "right").drop(columns = ["prominence"])

# Converting NaN to strings
tp_df["type"] = tp_df["type"].fillna("NA")

###############################
### Defining stable regions ###
###############################

# Selecting peaks above the mean line
significant_peaks = tp_df.loc[(tp_df["type"] == "peak") & (tp_df["sliding_window"] > mean_stability), "Pos"].tolist()

# Initialise variables
stabilisers = []
region_list = []
region = 1
stabiliser_count = 0

# Read min_stable_region_size argument
if len(sys.argv) > 2: 
    try: 
        min_region_size = int(sys.argv[2]) 
        print(f"Window size: {min_region_size}") 
    except ValueError: 
        print("Warning: invalid min region size provided — defaulting to 0") 
        min_region_size = 0 
else: 
    min_region_size = 0 
    print("No min region size provided, defaulting to 0")


# Convert to list for faster iteration
positions = mean_dG_by_pos["Pos"].tolist()
energies = mean_dG_by_pos["sliding_window"].tolist()

# Looping through each residue position
for i in range(len(positions)):
    pos = positions[i]
    energy = energies[i]

    # Check if the current position is stabilising
    if (energy < mean_stability):
        stabilisers.append(pos)
        region_list.append(region)
        stabiliser_count += 1

    # If a significant peak is hit
    if pos in significant_peaks:
        # Remove regions smaller than min_region_size
        if stabiliser_count < min_region_size:
            stabilisers = [s for s, r in zip(stabilisers, region_list) if r != region]
            region_list = [r for r in region_list if r != region]

        # Increment region for next potential stable region
        region += 1
        stabiliser_count = 0

# Handle the last region in case the sequence ends without a peak
if stabiliser_count < min_region_size:
    stabilisers = [s for s, r in zip(stabilisers, region_list) if r != region]
    region_list = [r for r in region_list if r != region]

# Create a DataFrame of stabilising positions and their assigned region
stabilising_positions = pd.DataFrame({
    "Pos": stabilisers,
    "region": region_list
})

# Get the unique region numbers, sorted in ascending order
unique_regions = sorted(stabilising_positions["region"].unique())

# Make sure regions start at 1 and increase by 1 integer between each region
region_map = {old: new for new, old in enumerate(unique_regions, start = 1)}
stabilising_positions["region"] = stabilising_positions["region"].map(region_map)

# Group stabilising positions by region and compute min and max Pos in each region
stable_regions = (
    stabilising_positions
    .groupby("region", as_index = False)
    .agg(min = ("Pos", "min"), max = ("Pos", "max"))
)

# Create a string describing the residue range in each region
stable_regions["residues"] = stable_regions["min"].astype(str) + "-" + stable_regions["max"].astype(str)

# Convert 'region' to a categorical type with sorted order
stable_regions["region"] = stable_regions["region"].astype(str)
# Sort the regions numerically, but keep as strings
numeric_sorted_regions = [str(i) for i in sorted(stable_regions["region"].astype(int))]
# Assign categorical with numeric order
stable_regions["region"] = pd.Categorical(
    stable_regions["region"],
    categories=numeric_sorted_regions,
    ordered=True
)

# Saving stable_regions
stable_regions_minimal = stable_regions[["region", "residues"]]
stable_regions_minimal.to_csv(os.path.join(output_dir, "stable_regions.csv"), index = False)

# Using matplotlib colour palette tab20
n_regions = 20
cmap = plt.get_cmap("tab20")  # qualitative palette
# get first n_regions colours
colours = [cmap(i) for i in range(n_regions)]
# convert to hex
colours = [mcolors.to_hex(c) for c in colours]

# Extend the color vector to match the length of `region` if needed
region_colours = np.tile(colours, int(np.ceil(len(stable_regions) / len(colours))))[:len(stable_regions)]

# Creating a new column to adjust the width of the rectangles in the plot
stable_regions["min_adj"] = stable_regions["min"] - 0.5
stable_regions["max_adj"] = stable_regions["max"] + 0.5

#######################################
### Visualising Stabilising Regions ###
#######################################

# Plotting turning points
p = (
    ggplot(tp_df, aes(x = "Pos", y = "sliding_window")) 
    + geom_rect(stable_regions, aes(xmin="min_adj", xmax="max_adj", ymin=-np.inf, ymax=np.inf, fill="region"),
        alpha=0.75, inherit_aes=False)
    + geom_point(aes(colour = "type"), size = 2) 
    + geom_line(size = 1) 
    + scale_x_continuous(breaks = range(0, 9999, 10))
    + scale_y_continuous(expand = (0.05, 0.05))
    + geom_hline(yintercept = mean_stability, linetype = "dashed", size = 0.75, colour = "black")
    + scale_fill_manual(values = region_colours)
    + scale_color_manual(values = {"pit": "red", "peak": "blue", "NA": "black"},
                         breaks = ["pit", "peak", "NA"],
                         labels = ["Pit", "Peak", "NA"]) 
    + labs(
        title = f"Turning Points: Sliding Window of {window_size}",
        y = r'$\mathbf{\Delta G^{\circ}\ per\ residue\ (kcal\cdot mol^{-1})}$',
        x = "Residue Number",
        colour = "Type",
        fill = "Region"
    )
    + theme_minimal()
    + theme(
        panel_border = element_rect(size = 1, colour = "black"),
        panel_grid_major = element_line(colour = "grey", size = 0.25, linetype = "dashed"),
        panel_grid_minor = element_line(colour = "grey", size = 0.25, linetype = "dashed"),
        plot_title = element_text(size = 14, colour = "black", face = "bold", ha = "center"),
        axis_title = element_text(size = 12, colour = "black", face = "bold"),
        axis_text = element_text(size = 10, colour = "black", face = "bold"),
        legend_title = element_text(size = 12, colour = "black", face = "bold"),
        legend_text = element_text(size = 10, colour = "black")
    )
)
# Saving turning points plot
p.save(os.path.join(output_dir, "turning_points_and_regions.png"), height = 5, width = 10, dpi = 300)

# Creating a group variable based on pdb_id to include line breaks if residue numbers are not continuous
mean_dG_per_res_by_PDB = mean_dG_per_res_by_PDB.sort_values(['pdb_id', 'Pos'])
mean_dG_per_res_by_PDB['segment'] = mean_dG_per_res_by_PDB.groupby('pdb_id')['Pos'].diff().ne(1).cumsum()
mean_dG_per_res_by_PDB['interaction'] = mean_dG_per_res_by_PDB['pdb_id'].astype(str) + "_" + mean_dG_per_res_by_PDB['segment'].astype(str)

# Plotting stable regions and sliding window data
p = (
    ggplot(mean_dG_per_res_by_PDB, aes(x = "Pos")) 
    + geom_rect(stable_regions, aes(xmin="min_adj", xmax="max_adj", ymin=-np.inf, ymax=np.inf, fill="region"),
        alpha=0.75, inherit_aes=False)
    + geom_hline(yintercept = 0, colour = "black", size = 1, linetype = "solid", alpha = 0.9) 
    + geom_line(aes(y = "sliding_window", group = "interaction"),
                size = 0.5, alpha = 0.25) 
    + scale_x_continuous(breaks = range(0, 9999, 10))
    + scale_y_continuous(expand = (0.05, 0.05))
    + scale_fill_manual(values = region_colours)
    + scale_color_manual(values = {"pit": "red", "peak": "blue", "NA": "black"},
                         breaks = ["pit", "peak", "NA"],
                         labels = ["Pit", "Peak", "NA"]) 
    + labs(
        title = f"Rolling Average Using Sliding Window of Size {window_size}",
        y = r'$\mathbf{\Delta G^{\circ}\ per\ residue\ (kcal\cdot mol^{-1})}$',
        x = "Residue Number",
        colour = "Type",
        fill = "Region"
    )
    + theme_minimal()
    + theme(
        panel_border = element_rect(size = 1, colour = "black"),
        panel_grid_major = element_line(colour = "grey", size = 0.25, linetype = "dashed"),
        panel_grid_minor = element_line(colour = "grey", size = 0.25, linetype = "dashed"),
        plot_title = element_text(size = 14, colour = "black", face = "bold", ha = "center"),
        axis_title = element_text(size = 12, colour = "black", face = "bold"),
        axis_text = element_text(size = 10, colour = "black", face = "bold"),
        legend_title = element_text(size = 12, colour = "black", face = "bold"),
        legend_text = element_text(size = 10, colour = "black")
    )
)
p.save(os.path.join(output_dir, "sliding_window_and_regions.png"), height = 5, width = 10, dpi = 300)

# Plotting stable regions and mean deltaG per residue for each PDB
p = (
    ggplot(mean_dG_per_res_by_PDB, aes(x = "Pos")) 
    + geom_rect(stable_regions, aes(xmin="min_adj", xmax="max_adj", ymin=-np.inf, ymax=np.inf, fill="region"),
        alpha=0.75, inherit_aes=False)
    + geom_hline(yintercept = 0, colour = "black", size = 1, linetype = "solid", alpha = 0.9) 
    + geom_line(aes(y = "mean_energy", group = "interaction"),
                size = 0.5, alpha = 0.25) 
    + scale_x_continuous(breaks = range(0, 9999, 10))
    + scale_y_continuous(expand = (0.05, 0.05))
    + scale_fill_manual(values = region_colours)
    + scale_color_manual(values = {"pit": "red", "peak": "blue", "NA": "black"},
                         breaks = ["pit", "peak", "NA"],
                         labels = ["Pit", "Peak", "NA"]) 
    + labs(
        title = "",
        y = r'$\mathbf{\Delta G^{\circ}\ per\ residue\ (kcal\cdot mol^{-1})}$',
        x = "Residue Number",
        colour = "Type",
        fill = "Region"
    )
    + theme_minimal()
    + theme(
        panel_border = element_rect(size = 1, colour = "black"),
        panel_grid_major = element_line(colour = "grey", size = 0.25, linetype = "dashed"),
        panel_grid_minor = element_line(colour = "grey", size = 0.25, linetype = "dashed"),
        plot_title = element_text(size = 14, colour = "black", face = "bold", ha = "center"),
        axis_title = element_text(size = 12, colour = "black", face = "bold"),
        axis_text = element_text(size = 10, colour = "black", face = "bold"),
        legend_title = element_text(size = 12, colour = "black", face = "bold"),
        legend_text = element_text(size = 10, colour = "black")
    )
)
p.save(os.path.join(output_dir, "mean_delta_G_and_regions.png"), height = 5, width = 10, dpi = 300)
